app/data_providers/dispatch.py

```diff
@@ -78,13 +85,26 @@ async def by_publisher(self, resolved_id: int, domain: str, count: int) -> List[
         except Item2ItemError:
             # fallback to random by the same publisher for syndicated right rail
             recs = await self.item_recommender.random_by_publisher(domain, count=100)
-        return self._sample(recs, count)
+            random.shuffle(recs)
+        return self._to_corpus_items(recs, count)
+
+    @staticmethod
+    def _unique_domains_first(recs: List[RelatedItem]) -> List[RelatedItem]:
+        seen_domains = set()
+        duplicates = []
+        uniques = []
+        for r in recs:
+            if r.domain not in seen_domains:
+                uniques.append(r)
+                seen_domains.add(r.domain)
+            else:
+                duplicates.append(r)
+        return uniques + duplicates
```

Comment:
Similar logic is already implemented in `spread_publishers` and `spread_topics`. Might be good to standardize one a single approach to spread based on a generic attribute.

For this PR, can you move this function to `app/rankers/algorithms.py` so it's easy to keep track of the rankers that we have?

Reply:
I agree, I thought about this, but the requirements, use cases and input/output data structures are quite different with the home pub spread. I can move it to algorithms, yes